libslide/src/bignum/mul.rs

```diff
@@ -0,0 +1,146 @@
+#![allow(clippy::suspicious_arithmetic_impl)]
+use crate::bignum::utils::convert_poly;
+use crate::bignum::utils::fft;
+use crate::bignum::utils::ifft;
+use crate::bignum::utils::ispowerof2;
+use crate::bignum::utils::recast_user_vec;
+use crate::bignum::utils::round_complex_vec;
+use crate::bignum::Bignum;
+use std::cmp::Ordering;
+use std::ops;
+
+impl ops::Mul for Bignum {
+    type Output = Bignum;
+
+    fn mul(self, rhs: Bignum) -> Bignum {
+        let is_neg: bool = self.is_neg ^ rhs.is_neg;
+        // total length of return vector is p+q+1
+        let total_length = self.dec.len() + self.int.len() + rhs.dec.len() + rhs.int.len();
+        // need self and rhs dec lengths for split later
+        let dec_len = self.dec.len() + rhs.dec.len();
+        // we need to make the vectors the same length at both the decimal and integer level.
+        let (mut lhs_dec, mut rhs_dec) = match self.dec.len().cmp(&rhs.dec.len()) {
+            Ordering::Greater => {
+                let mut rhs_dec = vec![0; self.dec.len()];
+                for (i, e) in rhs.dec.iter().enumerate() {
+                    rhs_dec[i] = *e
+                }
+                (self.dec, rhs_dec)
+            }
+            Ordering::Less => {
+                let mut lhs_dec = vec![0; rhs.dec.len()];
+                for (i, e) in self.dec.iter().enumerate() {
+                    lhs_dec[i] = *e
+                }
+                (lhs_dec, rhs.dec)
+            }
+            Ordering::Equal => (self.dec, rhs.dec),
+        };
+
+        let (mut lhs_int, mut rhs_int) = match self.int.len().cmp(&rhs.int.len()) {
+            Ordering::Greater => {
+                let mut rhs_int = vec![0; self.int.len()];
+                for (i, e) in rhs.int.iter().enumerate() {
+                    rhs_int[i] = *e
+                }
+                (self.int, rhs_int)
+            }
+            Ordering::Less => {
+                let mut lhs_int = vec![0; rhs.int.len()];
+                for (i, e) in self.int.iter().enumerate() {
+                    lhs_int[i] = *e
+                }
+                (lhs_int, rhs.int)
+            }
+            Ordering::Equal => (self.int, rhs.int),
+        };
+
+        // need to make vectors mod 2 for fft
```

Comment:
I think this could be better phrased. Maybe "vector lengths must be a power of 2 for fft"?